### PR TITLE
[WIP] FEATURE: Add tutorial enrolments

### PR DESCRIPTION
**Pull request by AkshatBajaj**
# Description

This PR will add tutorial enrolment......

Fixes # (issue)

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation if appropriate
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have created or extended unit tests to address my new additions
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

If you have any questions, please contact @macite or @jakerenzella.

### CODE REVIEW

File: app/models/tutorial.rb [hunk 1]
```diff
@@ -7,6 +7,7 @@ class Tutorial < ActiveRecord::Base
 
   has_many   :projects, dependent: :nullify # Students
   has_many   :groups, dependent: :nullify
+  has_many   :tutorial_enrolments
```
Comment (AkshatBajaj) [inline app/models/tutorial.rb]: Dependent?

File: app/models/tutorial_enrolment.rb [hunk 2]
```diff
@@ -0,0 +1,10 @@
+class TutorialEnrolment < ActiveRecord::Base
+  belongs_to :tutorial
+  belongs_to :project
+
+  validates :tutorial, presence: true
+  validates :project,  presence: true
+
+  # Always add a unique index to the DB to prevent new records from passing the validations when checked at the same time before being written
+  validates_uniqueness_of :tutorial, :scope => :project, message: 'already exists for the selected student'
```
Comment (macite) [inline app/models/tutorial_enrolment.rb]: Need to test uniqueness within stream

Comment (macite) [reply app/models/tutorial_enrolment.rb]: Can validate campus for each

File: app/api/tutorial_enrolments_api.rb [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+      tutorial = Tutorial.find(params[:tutorial_id])
+      unless authorise? current_user, tutorial.unit, :get_students
+        error!({ error: 'Not authorised to get enrolments for the selected tutorial' }, 403)
+      end
+
+      tutorial.tutorial_enrolments.find(params[:id])
+    end
+
+    desc 'Delete an enrolment in the tutorial'
+    delete '/tutorials/:tutorial_id/enrolments/:id' do
```
[the remark this answers is not in the document]
Comment (macite) [reply app/api/tutorial_enrolments_api.rb]: Remove tutorial_id and also `:id`

We need to think about what is best to map project here... ideally moving forward we want to remove project id from the front end as well... Maybe for now we leave it as is.

File: test/factories/projects_factory.rb [hunk 4]
```diff
@@ -0,0 +1,10 @@
+FactoryGirl.define do
+  factory :project do
+    unit
+    campus
```
Comment (macite) [inline test/factories/projects_factory.rb]: Also need the student in here... check any other relation as well.

File: test/models/tutorial_enrolment_model_test.rb [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+class TutorialEnrolmentModelTest < ActiveSupport::TestCase
+  def test_default_create
+    tutorial_enrolment = FactoryGirl.create(:tutorial_enrolment)
+    assert tutorial_enrolment.valid?
+  end
+
+  def test_specific_create
+    project = FactoryGirl.create(:project)
+    tutorial = FactoryGirl.create(:tutorial)
+    tutorial_enrolment = FactoryGirl.create(:tutorial_enrolment, project: project, tutorial: tutorial)
```
Comment (macite) [inline test/models/tutorial_enrolment_model_test.rb]: Shouldn't we be testing how "we" enrol projects in tutorials?

Then also test all of the above things that I mentioned as missing validations.

File: app/api/tutorial_enrolments_api.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (AkshatBajaj) [reply app/api/tutorial_enrolments_api.rb]: So this is `/units/:unit_id/tutorials/:tutorial_abbr/enrolments/:project_id`

Remove this student from that tutorial in this unit

File: test/factories/projects_factory.rb (hunk 4 above)
Comment (AkshatBajaj) [reply test/factories/projects_factory.rb]: Student user

File: app/models/project.rb [hunk 6]
```diff
@@ -682,8 +703,6 @@ def task_completion_csv
       portfolio_status,
       grade > 0 ? grade : '',
       grade_rationale,
-      tutorial ? tutorial.abbreviation : '',
-      main_tutor.name
```
Comment (AkshatBajaj) [inline app/models/project.rb]: @macite Any potential issues?

File: app/serializers/project_serializer.rb [hunk 7]
```diff
@@ -56,10 +56,6 @@ def student_id
     object.student.username
   end
 
-  def tutor_name
-    object.main_tutor.first_name unless object.main_tutor.nil?
-  end
-
```
Comment (AkshatBajaj) [inline app/serializers/project_serializer.rb]: @macite Do we need this now?

File: app/api/tasks_api.rb [hunk 8]
```diff
@@ -24,19 +24,18 @@ class TasksApi < Grape::API
         error!({ error: 'You do not have permission to read these task details' }, 403)
       end
 
+      # TODO (stream) Do we need tutorial here?
```
Comment (AkshatBajaj) [inline app/api/tasks_api.rb]: @macite Any suggestions?

File: app/models/unit.rb [hunk 9]
```diff
@@ -1617,17 +1599,18 @@ def tasks_as_hash(data)
   # Return all tasks from the database for this unit and given user
   #
   def get_all_tasks_for(user)
+    # TODO (stream) Do we need tutorial here?
```
Comment (AkshatBajaj) [inline app/models/unit.rb]: @macite Any suggestions?

File: app/models/unit.rb [hunk 10]
```diff
@@ -1669,11 +1652,12 @@ def tasks_for_task_inbox(user)
   #   task_def_id => { ... }
   #
   def task_status_stats
+    # TODO (stream) Do we need tutorial here?
```
Comment (AkshatBajaj) [inline app/models/unit.rb]: @macite any suggestions?

File: app/api/tasks_api.rb (hunk 8 above)
Comment (macite) [reply app/api/tasks_api.rb]: Yes, we need tutorial_id here.
We should add stream_id as well.

File: app/models/project.rb (hunk 6 above)
Comment (macite) [reply app/models/project.rb]: Yes we will need the list of tutorials for each stream here.

Also need to adjust the header to include the stream abbreviations for each of these new columns.
Also need one column if there are no streams

File: app/models/unit.rb (hunk 9 above)
Comment (macite) [reply app/models/unit.rb]: Here we need the relevant tutorial id through the associated stream

File: app/models/unit.rb [hunk 11]
```diff
@@ -1669,11 +1652,12 @@ def tasks_for_task_inbox(user)
...
     data = student_tasks
            .joins(:task_status)
-           .select('projects.tutorial_id as tutorial_id', 'task_definition_id', 'task_statuses.id as status_id', 'COUNT(tasks.id) as num_tasks')
+           .select('task_definition_id', 'task_statuses.id as status_id', 'COUNT(tasks.id) as num_tasks')
            .where('task_status_id > 1')
-           .group('projects.tutorial_id', 'tasks.task_definition_id', 'status_id')
+           .group('tasks.task_definition_id', 'status_id')
            .map do |r|
       {
         tutorial_id: r.tutorial_id,
```
Comment (macite) [inline app/models/unit.rb]: See you need the tutorial id here...

File: app/models/unit.rb (hunk 10 above)
Comment (macite) [reply app/models/unit.rb]: Yes -> see the comment. Tutorial is essential here.

File: app/serializers/project_serializer.rb (hunk 7 above)
Comment (macite) [reply app/serializers/project_serializer.rb]: No - we have the tutorial enrolments... and we should be getting the name from the unit data.

File: app/serializers/tutorial_serializer.rb [hunk 12]
```diff
@@ -4,7 +4,12 @@
 require 'user_serializer'
 
 class TutorialSerializer < ActiveModel::Serializer
-  attributes :id, :meeting_day, :meeting_time, :meeting_location, :abbreviation, :campus_id, :capacity, :tutor_name, :num_students
+  attributes :id, :meeting_day, :meeting_time, :meeting_location, :abbreviation, :campus_id, :capacity, :tutor_name, :num_students,
```
Comment (macite) [inline app/serializers/tutorial_serializer.rb]: We can remove tutor name from here... we should be able to link to this if we have tutor id or something similar.

File: app/models/unit.rb [hunk 13]
```diff
@@ -392,42 +408,25 @@ def employ_staff(user, role)
...
 
   # Adds a user to this project.
-  def enrol_student(user, campus, tutorial = nil)
-    tutorial_id = if tutorial.is_a?(Tutorial)
-                    tutorial.id
-                  else
-                    tutorial
-                  end
-
+  def enrol_student(user, campus)
```
Comment (AkshatBajaj) [inline app/models/unit.rb]: Remove `tutorial` from method call (line 717)

Comment (macite) [reply app/models/unit.rb]: Yes, I think we will have to now. Where was this used?

**Pull request merged**